### PR TITLE
[query] Adds histogram_quantile function

### PR DESCRIPTION
**Pull request by arnikola**

### CODE REVIEW

File: src/query/functions/linear/histogram_quantile.go [hunk 1]
```diff
@@ -0,0 +1,376 @@
...
+func NewHistogramQuantileOp(
+	args []interface{},
+	opType string,
+) (parser.Params, error) {
+	if len(args) != 1 {
+		return emptyOp, fmt.Errorf(
+			"invalid number of args for histogram_quantile: %d", len(args))
+	}
+
+	if opType != HistogramQuantileType {
```
Comment (benraskin92) [inline src/query/functions/linear/histogram_quantile.go]: Check this first

Comment (arnikola) [reply src/query/functions/linear/histogram_quantile.go]: Was copying the pattern [from clamp.go](https://github.com/m3db/m3/blob/master/src/query/functions/linear/clamp.go#L49)

Wouldn't mind dealing with these correctly within the parser in a later pr, like the binary and aggregation functions are done at the moment.

File: src/query/functions/linear/histogram_quantile_test.go [hunk 2]
```diff
@@ -0,0 +1,283 @@
...
+		{upperBound: 1, value: 1},
+		{upperBound: 2, value: 2},
+	})
+	assert.True(t, math.IsNaN(actual))
+
+	actual = bucketQuantile(0.5, []bucketValue{
+		{upperBound: 1, value: 1},
+		{upperBound: math.Inf(1), value: 22},
+	})
+	assert.Equal(t, float64(1), actual)
```
Comment (benraskin92) [inline src/query/functions/linear/histogram_quantile_test.go]: Add test with neg infinity and > 2 buckets?

File: src/query/functions/linear/histogram_quantile_test.go [hunk 3]
```diff
@@ -0,0 +1,283 @@
...
+
+	return sink.Values
+}
+
+var (
+	inf  = math.Inf(+1)
+	ninf = math.Inf(-1)
+)
+
+func TestQuantileFunctionFilteringWithoutA(t *testing.T) {
```
Comment (benraskin92) [inline src/query/functions/linear/histogram_quantile_test.go]: `WithoutA`?

Comment (benraskin92) [reply src/query/functions/linear/histogram_quantile_test.go]: Same with `WithQ`?

File: src/query/functions/linear/histogram_quantile_test.go [hunk 4]
```diff
@@ -0,0 +1,283 @@
...
+	op, err := NewHistogramQuantileOp(args, HistogramQuantileType)
+	assert.NoError(t, err)
+
+	assert.Equal(t, HistogramQuantileType, op.OpType())
+	assert.Equal(t, "type: histogram_quantile", op.String())
+}
+
+func testQuantileFunctionWithQ(t *testing.T, q float64) [][]float64 {
+	args := make([]interface{}, 0, 1)
+	args = append(args, q)
```
Comment (benraskin92) [inline src/query/functions/linear/histogram_quantile_test.go]: Just pass in the args instead of having to construct them here?

File: src/query/functions/linear/histogram_quantile.go [hunk 5]
```diff
@@ -0,0 +1,376 @@
...
+		return err
+	}
+
+	meta := stepIter.Meta()
+	seriesMetas := utils.FlattenMetadata(meta, stepIter.SeriesMeta())
+	bucketedSeries := gatherSeriesToBuckets(seriesMetas)
+
+	q := n.op.q
+	if q < 0 || q > 1 {
+		return processInvalidQuantile(q, bucketedSeries, meta, stepIter, n.controller)
```
Comment (benraskin92) [inline src/query/functions/linear/histogram_quantile.go]: Seems kind of weird to split this out like this. Maybe just have one function called `processQuantile` and put the inf logic in there.

File: src/query/functions/linear/histogram_quantile.go [hunk 6]
```diff
@@ -0,0 +1,376 @@
...
+)
+
+const (
+	// HistogramQuantileType calculates the quantile for histogram buckets.
+	//
+	// NB: each sample must contain a tag with a bucket name (given by tag
+	// options) that denotes the upper bound of that bucket; series without this
+	// tag are ignored.
+	HistogramQuantileType = "histogram_quantile"
+	initIndexBucketLength = 10
```
Comment (benraskin92) [inline src/query/functions/linear/histogram_quantile.go]: Any specific reason to use 10?

File: src/query/functions/linear/histogram_quantile.go [hunk 7]
```diff
@@ -0,0 +1,376 @@
...
+	if err != nil {
+		return err
+	}
+
+	for index := 0; stepIter.Next(); index++ {
+		step := stepIter.Current()
+		values := step.Values()
+		bucketValues := make([]bucketValue, 0, initIndexBucketLength)
+
+		aggregatedValues := make([]float64, 0, len(bucketedSeries))
```
Comment (benraskin92) [inline src/query/functions/linear/histogram_quantile.go]: You should be able to just pre-allocate this with a length and index in below when adding to it. Looks like you always add to it.

File: src/query/models/tags_test.go [hunk 8]
```diff
@@ -226,18 +226,56 @@ func TestUpdateName(t *testing.T) {
...
+	value := []byte("n")
+	tags = tags.SetBucket(value)
+	actual, found = tags.Get(name)
+	assert.True(t, found)
+	assert.Equal(t, value, actual)
+	actual, found = tags.Bucket()
+	assert.True(t, found)
+	assert.Equal(t, value, actual)
+
+	value2 := []byte("abc")
```
Comment (benraskin92) [inline src/query/models/tags_test.go]: Add test for empty val? e.g. `[]byte("")`

File: src/cmd/services/m3query/config/config.go [hunk 9]
```diff
@@ -299,6 +299,10 @@ type TagOptionsConfiguration struct {
 	// If not provided, defaults to `__name__`.
 	MetricName string `yaml:"metricName"`
 
+	// BucketName specifies the tag name that corresponds to the metric's bucket.
+	// If not provided, defaults to `le`.
```
[the remark this answers is not in the document]
Comment (benraskin92) [reply src/cmd/services/m3query/config/config.go]: Nvm

File: src/query/functions/linear/histogram_quantile.go [hunk 10]
```diff
@@ -0,0 +1,376 @@
...
+		sign = -1
+	}
+
+	setValue := math.Inf(sign)
+	outValues := make([]float64, len(bucketedSeries))
+	ts.Memset(outValues, setValue)
+
+	for index := 0; stepIter.Next(); index++ {
+		cloned := make([]float64, len(outValues))
+		copy(cloned, outValues)
```
[the remark this answers is not in the document]
Comment (arnikola) [reply src/query/functions/linear/histogram_quantile.go]: Nice catch, copied from the regular path, forgot to fix this one

File: src/query/functions/linear/histogram_quantile.go [hunk 11]
```diff
@@ -0,0 +1,376 @@
...
+func processValidQuantile(
+	q float64,
+	bucketedSeries bucketedSeries,
+	meta block.Metadata,
+	stepIter block.StepIter,
+	controller *transform.Controller,
+) error {
+	sanitizeBuckets(bucketedSeries)
+
+	builder, err := setupBuilder(bucketedSeries, meta, stepIter, controller)
```
[the remark this answers is not in the document]
Comment (arnikola) [reply src/query/functions/linear/histogram_quantile.go]: Need to set a bunch of meta, seriesMetas, step counts unfortunately

File: src/cmd/services/m3query/config/config.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (arnikola) [reply src/cmd/services/m3query/config/config.go]: Yep :+1:

File: src/query/functions/linear/histogram_quantile_test.go (hunk 2 above)
Comment (arnikola) [reply src/query/functions/linear/histogram_quantile_test.go]: Good idea

File: src/query/functions/linear/histogram_quantile_test.go (hunk 3 above)
Comment (arnikola) [reply src/query/functions/linear/histogram_quantile_test.go]: Oh I copied some setup stuff from an agg function that filtered by an `A` tag, will update this name to make more sense

File: src/query/functions/linear/histogram_quantile_test.go (hunk 4 above)
Comment (arnikola) [reply src/query/functions/linear/histogram_quantile_test.go]: Unfortunately it's pretty annoying to build a `[]interface{..}`, and that's what we need to pass into the function; since we only ever need the `q` value here, it should be fine to build the interface slice here

File: src/query/models/tags_test.go (hunk 8 above)
Comment (arnikola) [reply src/query/models/tags_test.go]: Sure; not sure if it's necessary a special case though

File: src/query/functions/linear/histogram_quantile.go (hunk 7 above)
Comment (arnikola) [reply src/query/functions/linear/histogram_quantile.go]: Will do; was fine using `append` here since didn't want to add a tracking `idx` variable

File: src/query/functions/linear/histogram_quantile.go (hunk 5 above)
Comment (arnikola) [reply src/query/functions/linear/histogram_quantile.go]: If we do that, will need to do the comparison for each step

File: src/query/functions/linear/histogram_quantile.go (hunk 6 above)
Comment (arnikola) [reply src/query/functions/linear/histogram_quantile.go]: Seems to be the general convention for initing arrays we don't know the length of

Here it has (by chance) a bit more context in that the general rule of thumb for histograms is no more than 10 buckets

File: src/query/functions/linear/histogram_quantile.go [hunk 12]
```diff
@@ -0,0 +1,375 @@
...
+}
+
+func setupBuilder(
+	bucketedSeries bucketedSeries,
+	meta block.Metadata,
+	stepIter block.StepIter,
+	controller *transform.Controller,
+) (block.Builder, error) {
+	metas := make([]block.SeriesMeta, len(bucketedSeries))
+	idx := 0
```
Comment (benraskin92) [inline src/query/functions/linear/histogram_quantile.go]: Why have this var? Just include `idx` in the for loop below?

File: src/query/functions/linear/histogram_quantile_test.go [hunk 13]
```diff
@@ -0,0 +1,291 @@
...
+)
+
+func TestQuantileFunctionForInvalidQValues(t *testing.T) {
+	actual := testQuantileFunctionWithQ(t, -1)
+	assert.Equal(t, [][]float64{{ninf, ninf, ninf, ninf, ninf}}, actual)
+	actual = testQuantileFunctionWithQ(t, 1.1)
+	assert.Equal(t, [][]float64{{inf, inf, inf, inf, inf}}, actual)
+
+	actual = testQuantileFunctionWithQ(t, 0.8)
+	test.EqualsWithNansWithDelta(t, [][]float64{{15.6, 20, 11.6, 13.4, 0.8}}, actual, 0.00001)
```
Comment (benraskin92) [inline src/query/functions/linear/histogram_quantile_test.go]: Add a case with some `NaNs`

File: src/query/functions/linear/histogram_quantile.go (hunk 12 above)
Comment (arnikola) [reply src/query/functions/linear/histogram_quantile.go]: It's a map not an array

File: src/query/functions/linear/histogram_quantile.go [hunk 14]
```diff
@@ -0,0 +1,375 @@
...
+			continue
+		}
+
+		excludeTags := [][]byte{tags.Opts.MetricName(), tags.Opts.BucketName()}
+		tagsWithoutKeys := tags.TagsWithoutKeys(excludeTags)
+		id := tagsWithoutKeys.ID()
+		if buckets, found := bucketsForID[string(id)]; !found {
+			// Add a single indexed bucket for this ID with the current index only.
+			newBuckets := make([]indexedBucket, 0, initIndexBucketLength)
+			newBucket := indexedBucket{
```
[the remark this answers is not in the document]
Comment (arnikola) [reply src/query/functions/linear/histogram_quantile.go]: Good call; this was originally going to have a list of indices that are valid for that bucket but the logic got really awkward. Changed it to add a bucket/index combo every time instead, but didn't update here to reflect that

File: src/query/functions/linear/histogram_quantile.go (hunk 12 above)
Comment (benraskin92) [reply src/query/functions/linear/histogram_quantile.go]: 🤦‍♂️
